src/AssociativeOpsTable.h

```diff
@@ -0,0 +1,63 @@
+#ifndef HALIDE_ASSOCIATIVE_OPS_TABLE_H
+#define HALIDE_ASSOCIATIVE_OPS_TABLE_H
+
+/** \file
+ * Tables listing associative operators and their identities.
+ */
+
+#include "IROperator.h"
+#include "IREquality.h"
+
+#include <iostream>
+#include <vector>
+
+namespace Halide {
+namespace Internal {
+
+// Represent an associative op with its identity. The op may be multi-dimensional,
+// e.g. complex multiplication. 'is_commutative' is set to true if the op is also
+// commutative in addition to being associative.
+struct AssociativePattern {
+    std::vector<Expr> ops;
+    std::vector<Expr> identities;
+    bool is_commutative;
+
+    AssociativePattern() : is_commutative(false) {}
```

Comment:
Need docstrings on all of these methods. Consider moving implementations into AssociativeOpsTable.cpp too - I don't think debugging print code belongs in a header (maybe that should go in IRPrinter with the rest of our operator<< overloads)